feat: namespace as a runtime routing dimension (engine + SDKs) (#1984)
MotiaDev/motia: docs/next/upgrading/from-0-22-x.mdx

---
title: "Upgrading from 0.22.x to 0.23.x"
description:
  "Redeclare RabbitMQ durable subscriber queues, then adopt namespaces on the surfaces your project
  uses."
owner: "devrel"
type: "how-to"
---

0.23.x adds namespaces as a routing dimension. Workers that declare no namespace land in `default`
and behave exactly as before, so most projects upgrade with no code change. The one hard break is
durable queue naming on the RabbitMQ adapter. Apply the steps below that touch surfaces your project
uses.

## Step 1: Redeclare RabbitMQ durable subscriber queues

Durable subscriber queue names are now namespace-qualified, so two subscribers of the same topic and
function id in different namespaces get two queues instead of competing for one. The RabbitMQ adapter
has no automatic migration: queues declared by a pre-0.23 deploy keep their old names and stop being

> addressed after the upgrade.

Reviewer: I'm unclear what `queues declared by a pre-0.23 deploy keep their old names and stop being
addressed after the upgrade.` means. Is this saying that old named queues stop being used after upgrade? Can make this clearer if so.

Author: Addressed in commit [`7a698ef3`](https://github.com/iii-hq/iii/commit/7a698ef3). The guide now states that a 0.23 subscriber declares and consumes a new namespace-qualified queue, does not rename, consume, or migrate the old queue, and that the old durable binding can continue to route copies until the queue is deleted.

| Version | Subscriber queue                        | Dead-letter queue                     |
| ------- | --------------------------------------- | ------------------------------------- |
| 0.22.x  | `iii.{topic}.{function_id}.queue`       | `iii.{topic}.{function_id}.dlq`       |
| 0.23.x  | `iii.{topic}.{function_id}@{ns}.queue`  | `iii.{topic}.{function_id}@{ns}.dlq`  |

`{ns}` is the subscriber's namespace, `default` when it declared none. Any `@` or `\` inside a
topic, function id, or namespace is backslash-escaped before the join, so the single unescaped `@` is

> always the namespace boundary.

Reviewer: ```suggestion
topic, function id, or namespace is backslash-escaped before the join.
```

<Warning>
  Drain the old queues before upgrading. Messages left in a pre-0.23 queue are not migrated and no
  consumer will read them afterwards.
</Warning>

For each affected subscriber:

1. Stop publishing to the topic.
2. Let the existing consumers drain the old queue to empty.
3. Upgrade and restart. The adapter declares the new namespace-qualified queue on the next subscribe.
4. Delete the drained queue and its dead-letter queue from the broker.

## Step 2: Verify the builtin broker migrated

> The builtin broker migrates itself. On the first boot after the upgrade it drains each legacy

Reviewer: What is meant by builtin broker? The engine's queue broker?
Also we should have no notion of builtin in the docs.
```suggestion
The engine's internal broker migrates itself. On the first boot after the upgrade it drains each legacy
```

Author: Addressed in commit [`3da9b5b4`](https://github.com/iii-hq/iii/commit/3da9b5b4). The upgrade guide now uses `engine-managed queue storage` and explains the startup migration sequence and its limits without using `builtin broker`.

subscriber queue into its namespaced identity and logs one line per queue:

```text
Drained legacy subscriber queue into its namespace
```

No action is needed. A queue name that already contains the separator is skipped and logged as such,
because it cannot be told apart from an already-migrated name.

> ## Step 3: Tune the registration grace, if needed

Reviewer: ```suggestion
## Step 3: Tune the registration grace period, if needed
```

A connection's namespace arrives on its `engine::workers::register` call, which SDKs send after their
first registration frames. The engine buffers a connection's registrations until the namespace is
known, or until a grace period expires and the connection is fixed to `default`.

> The default grace is 5000 ms. Raise it if workers on slow links register but land in `default`

Reviewer: ```suggestion
The default grace period is 5000 ms. Raise it if workers on slow links register but land in `default`
```

unexpectedly:

```yaml config.yaml
workers:
  - name: iii-worker-manager
    config:

>       registration_namespace_grace_ms: 10000

Reviewer: I'm surprised max grace period is a per-worker config and not an engine-wide config. We should reconsider this. Also this means workers must already be known to the engine and have a configuration before their grace period can be adjusted which is awkward to do in a system with live registration.

Author: Good point. The grace period is now an engine-wide setting at the root of config.yaml. The engine loads it before it creates any workers, so live workers do not need to be known or configured in advance. Resolved in commit [a0e3af6e](https://github.com/iii-hq/iii/commit/a0e3af6ee8339c2c9cc1289f86514ff2fca399e4).

> ```

Reviewer: This needs to be documented in using-iii as well

Author: Addressed in commit [`b1b178b8`](https://github.com/iii-hq/iii/commit/b1b178b8). The `using-iii/configuration` guide now documents the global `registration_namespace_grace_ms` setting, its `5000 ms` default, and the engine-process `III_NAMESPACE_GRACE_MS` override.

> `III_NAMESPACE_GRACE_MS` overrides the configured value at runtime and takes precedence.

Reviewer: Where is this set? engine's environment or the worker's?

Author: Good question. `III_NAMESPACE_GRACE_MS` is read from the engine process environment, not from a worker environment. It applies to namespace resolution for all new worker connections and overrides `registration_namespace_grace_ms` from `config.yaml`. Clarified in commit [`4bd886f4`](https://github.com/iii-hq/iii/commit/4bd886f4) in both the upgrade guide and the configuration documentation.

## Step 4: Check for `engine::*` function ids outside `default`

> 

Reviewer: ```suggestion
<Note>
Your own workers are unlikely to have `engine::` prefixed function names but we've documented this as it would be a breaking change if they do exist.
</Note>
```

> `engine::*` is reserved for engine and builtin infrastructure, which lives in `default`. A worker

Reviewer: ```suggestion
`engine::*` is reserved for engine and builtin infrastructure, which is in `default`. A worker
```

Author: The current wording uses `engine infrastructure` and does not expose the internal `builtin` term. This also matches the earlier feedback on the queue migration documentation, so I kept the more specific current wording.

that registers an `engine::*` function id while running in another namespace now has that one
registration refused with a `FUNCTION_NAMESPACE_CONFLICT`; the connection stays open and its other
functions keep serving. Rename any such id to your own `service::name` prefix.

This only affects workers that both use a namespace and register under the `engine::` prefix, so a

> project that has not adopted namespaces is unaffected.

Reviewer: Does every project not inherently use namespaces going forward? Even those that don't specify will get the default namespace.
```suggestion
```

Author: You are right. Every worker has an effective namespace. When none is specified, the worker uses the default namespace. The intended distinction here is between the default namespace and a non-default namespace, not between projects that have or have not adopted namespaces. I updated the wording to distinguish non-default namespaces and removed the final paragraph in commit [3a17f333](https://github.com/iii-hq/iii/commit/3a17f333fdf0fc0450c8c88a0a25ab030bdb1953).

## Step 5: Scope RBAC expose rules when adopting a namespace

`expose_functions` rules on the `iii-worker-manager` RBAC listener are namespace-scoped. A rule that
names no namespace applies to `default` only, so it will not expose a function reached in another
namespace. When you move a worker into a namespace, scope its rules to match:

```yaml
expose_functions:
  - match: "orders::*"

>     namespace: orders

Reviewer: Do namespaces accept glob matches in RBAC workers? I believe we discussed allowing it, if so we should show a second example here of that.

Author: Addressed in commit [`1e0425b`](https://github.com/iii-hq/iii/commit/1e0425beb3e4531c1fd54df319a0fa47786a6e87). Namespace values do not support glob matching in the current implementation. Function-id `match` values support `*`, while `namespace` uses an exact match. The worker-manager and upgrade guides now state this and require one rule per namespace.

```

`forbidden_functions` and `allowed_functions` still match on the bare function id and apply in every
namespace; they come from the session's auth result, which carries no namespace.

> 

Reviewer: This needs to be documented in creating-workers/worker-manager

Author: Addressed in commit [`1e0425b`](https://github.com/iii-hq/iii/commit/1e0425beb3e4531c1fd54df319a0fa47786a6e87). The worker-manager guide now has a “Namespace-scoped function rules” section with function-id and metadata examples. It also explains the `default` scope and exact namespace matching.

| 0.23.x  | `iii.{topic}.{function_id}@{ns}.queue`  | `iii.{topic}.{function_id}@{ns}.dlq`  |

`{ns}` is the subscriber's namespace, `default` when it declared none. Any `@` or `\` inside a
topic, function id, or namespace is backslash-escaped before the join, so the single unescaped `@` is
always the namespace boundary.

<Warning>
  Drain the old queues before upgrading. Messages left in a pre-0.23 queue are not migrated and no
  consumer will read them afterwards.
</Warning>

For each affected subscriber:

1. Stop publishing to the topic.
2. Let the existing consumers drain the old queue to empty.
3. Upgrade and restart. The adapter declares the new namespace-qualified queue on the next subscribe.
4. Delete the drained queue and its dead-letter queue from the broker.

## Step 2: Verify the builtin broker migrated

The builtin broker migrates itself. On the first boot after the upgrade it drains each legacy
subscriber queue into its namespaced identity and logs one line per queue:

```text
Drained legacy subscriber queue into its namespace
```

No action is needed. A queue name that already contains the separator is skipped and logged as such,

> because it cannot be told apart from an already-migrated name.

Author: Addressed in commit [`1e0425b`](https://github.com/iii-hq/iii/commit/1e0425beb3e4531c1fd54df319a0fa47786a6e87). The migration checklist now requires recovery of every legacy queue skipped because its name contains `@`. The existing warning gives the recovery sequence: stop publishing, restart 0.22.x with the same storage, drain the queue, and retry the upgrade.
